pilot/pkg/controllers/ipallocate/ipallocate.go

```diff
@@ -0,0 +1,402 @@
+	return allocator
+}
+
+func (c *IPAllocate) Run(stop <-chan struct{}) {
+	log.Debugf("starting %s controller", controllerName)
+	kubelib.WaitForCacheSync(controllerName, stop, c.serviceEntryClient.HasSynced)
+	// it is important that we always warm cache before we try to run the queue
+	// failing to warm cache first could result in allocating IP which are already in use
+	c.warmCache()
+	// logically the v1 behavior is a state of the world function which cannot be replicated in reconcile
+	// if we wish to replicate v1 assignment behavior for existing serviceentry it should go here and must
+	// include a change to respect the used IPs from the warmed cache begin reconcile
+	c.queue.Run(stop)
+	c.serviceEntryClient.ShutdownHandlers()
+}
+
+// warmCache reads all serviceentries and records any auto-assigned addresses as in use
+func (c *IPAllocate) warmCache() {
+	serviceentries := c.serviceEntryClient.List(metav1.NamespaceAll, klabels.Everything())
+	count := 0
+	for _, serviceentry := range serviceentries {
+		count++
+		owner := types.NamespacedName{
+			Name:      serviceentry.Name,
+			Namespace: serviceentry.Namespace,
+		}
+		log.Debugf("%s/%s found during warming", serviceentry.Namespace, serviceentry.Name)
+		if len(serviceentry.Spec.Addresses) > 0 {
+			for _, addr := range serviceentry.Spec.Addresses {
+				a, err := netip.ParseAddr(addr)
+				if err != nil {
+					log.Debugf("unable to parse address %s for %s/%s, received error: %s", addr, serviceentry.Namespace, serviceentry.Name, err.Error())
+					continue
+				}
+				// these are not assigned by us but could conflict with our IP ranges and cause issues
+				if !c.inOurRange(a) {
+					// don't need to worry about this one
+					continue
+				}
+				c.markUsedOrQueueConflict(a, owner)
+			}
+		}
+
+		if serviceentry.Status.String() != "" {
+			addresses := autoallocate.GetV2AddressesFromServiceEntry(serviceentry)
+			if len(addresses) > 0 {
+				for _, a := range addresses {
+					c.markUsedOrQueueConflict(a, owner)
+				}
+			}
+		}
+	}
+	log.Debugf("discovered %v during warming", count)
+}
+
+func (c *IPAllocate) reconcile(a any) error {
+	if nn, ok := a.(types.NamespacedName); ok {
+		return c.reconcileServiceEntry(nn)
+	}
+	if conflict, ok := a.(conflictDetectedEvent); ok {
+		return c.resolveConflict(conflict)
+	}
+
+	return nil
+}
+
+func (c *IPAllocate) reconcileServiceEntry(se types.NamespacedName) error {
+	log.Debugf("reconciling ServiceEntry %s/%s", se.Namespace, se.Name)
+	serviceentry := c.serviceEntryClient.Get(se.Name, se.Namespace)
+	if serviceentry == nil {
+		// probably a delete so we should remove ips from our addresses most likely
+		// TODO: we never actually remove IP right now, likely this should be done a little more slowly anyway to prevent reuse if we are too fast
+		return nil
+	}
+
+	// TODO: also what do we do if we already allocated IPs in a previous reconcile but are not longer meant to
```

Comment:
I think we will need to remove the address from the status?

Reply:
Probably so, we could do so without freeing in the data structure to prevent immediate reuse and it would get gc once that impl is worked out